pkg/retriever/assignablecandidatefilter.go

```diff
@@ -0,0 +1,113 @@
+package retriever
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/benbjohnson/clock"
+	"github.com/filecoin-project/lassie/pkg/events"
+	"github.com/filecoin-project/lassie/pkg/internal/stream"
+	"github.com/filecoin-project/lassie/pkg/types"
+)
+
+// AssignableCandidateFinder finds and filters candidates for a given retrieval
+type AssignableCandidateFinder struct {
+	candidateFinder             CandidateFinder
+	isAcceptableStorageProvider IsAcceptableStorageProvider
+}
+
+func NewAssignableCandidateFinder(candidateFinder CandidateFinder, isAcceptableStorageProvider IsAcceptableStorageProvider) types.CandidateFinder {
+	return &AssignableCandidateFinder{candidateFinder: candidateFinder, isAcceptableStorageProvider: isAcceptableStorageProvider}
+}
+
+func (rcf *AssignableCandidateFinder) FindCandidates(ctx context.Context, request types.RetrievalRequest, eventsCallback func(types.RetrievalEvent)) (types.CandidateStream, error, <-chan error) {
+	ctx, cancelCtx := context.WithCancel(ctx)
+	defer cancelCtx()
+	phaseStarted := time.Now()
+
+	eventsCallback(events.Started(request.RetrievalID, phaseStarted, types.IndexerPhase, types.RetrievalCandidate{RootCid: request.Cid}))
+
+	// start the call to get candidates
+	individualCandidateChan, err := rcf.candidateFinder.FindCandidatesAsync(ctx, request.Cid)
+
+	// if we get an immediate error, stop
+	if err != nil {
+		eventsCallback(events.Failed(request.RetrievalID, phaseStarted, types.IndexerPhase, types.RetrievalCandidate{RootCid: request.Cid}, err.Error()))
+		return nil, fmt.Errorf("could not get retrieval candidates for %s: %w", request.Cid, err), nil
+	}
+
+	// convert our candidate result channel to a stream
+	individualCandidateStream := stream.NewResultChannelStream(ctx, individualCandidateChan)
```

Comment:
ok, but how do you deal with deduping candidates and the potential for having to prioritise `FastRetrieval` and the possibility of having duplicates with and without `FastRetrieval`?
described in more detail @ https://github.com/filecoin-project/lassie/pull/71#issuecomment-1425243875

Reply:
It's for sure doable -- just write a stateful map

```golang
func distinctCandidates(incomingStream types.CandidateStream, candidateKey func(RetrievalCanddate) interface{}) types.CandidateStream {
   candidateSet := make(map[interface{}]struct{})
   dedupedStream := stream.Map(incomingStream, func(nextCandidates []types.RetrievalCandidate) []types.RetrievalCandidate {
       filteredCandidates := make([]types.RetrievalCandidate, 0, len(nextCandidates))
       for _, nextCandidate := range nextCandidates {
          key := candidateKey(nextCandidate)
          if _, ok := candidateSet[key]; ok {
              continue
          }
          candidateSet[key] = struct{}{}
          filteredCandidates = append(filteredCandidates, nextCandidate)
       }
       return filteredCandidates
   })
   return stream.Filter(dedupedStream, func(nextCandidates []types.RetrievalCandidate) bool { return len(nextCandidates) > 0})
}
```


I have no idea if that code is right and of course candidateKey is the critical bit of functionality but am I missing something?